src/xlat-i/xi-Parameter.C

```diff
@@ -619,18 +756,29 @@ void Parameter::beginUnmarshallArray(XStr& str) {
   }
 }
 
-void Parameter::beginUnmarshallRdma(XStr& str,
-                                    bool genRdma) {  // First pass: unpack pup'd entries
-  Type* dt = type->deref();                          // Type, without &
+// First pass: unpack pup'd entries
+void Parameter::beginUnmarshallRdma(XStr& str, bool genRdma, bool device) {
+  Type* dt = type->deref(); // Type, without &
+
   if (isRdma()) {
-    if (genRdma) {
+    bool hostPath = !device && !isDevice();
+    bool devicePath = device && isDevice();
+
+    if (hostPath && genRdma) {
       str << "  CkNcpyBuffer ncpyBuffer_" << name << ";\n";
       str << "  implP|ncpyBuffer_" << name << ";\n";
 
-      str << "  " << dt << " *ncpyBuffer_" << name <<"_ptr = ";
-      str << "("<<dt<< " *)" << " ncpyBuffer_" << name <<".ptr;\n";
-    } else
+      str << "  " << dt << " *ncpyBuffer_" << name << "_ptr = ";
+      str << "(" << dt << " *)" << " ncpyBuffer_" << name << ".ptr;\n";
+    } else if (devicePath) {
+      str << "  CkDeviceBuffer deviceBuffer_" << name << ";\n";
+      str << "  implP|deviceBuffer_" << name << ";\n";
+
+      str << "  " << dt << " *deviceBuffer_" << name << "_ptr = ";
+      str << "(" << dt << " *)" << " deviceBuffer_" << name << ".ptr;\n";
+    } else if (hostPath && !genRdma) {
```

Comment:
For better readability, you can just have 
```
if(hostPath) {
   if(genRdma) {
       ...
   } else {
       ...
   }
} else if(devicePath) {
    ....
}
```

This will keep the cpu code together.

Reply:
👍